File: Algorithms/1209_Remove_All_Adjacent_Duplicates_in_String_II/Java/RemoveAllAdjacentDuplicatesInStringIiSolution2.java

// Space: O(n)
// Time: O(n)

class Solution {
    public String removeDuplicates(String s, int k) {
        if (s.length() <= 1) return s;
        if (s.length() < k) return s;

        Stack<Pair<Character, Integer>> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (stack.size() == 0 || (stack.size() > 0 && stack.peek().getKey() != c)) {
                Pair<Character, Integer> newItem = new Pair<>(c, 1);
                stack.add(newItem);
            } else if (stack.peek().getKey() == c) {
                Pair<Character, Integer> top = stack.pop();
                int count = top.getValue();
                count++;
                if (count < k) {
                    stack.add(new Pair<Character, Integer>(top.getKey(), count));
                }
            }
        }

        StringBuilder res = new StringBuilder();
        while (stack.size() > 0) {
            Pair<Character, Integer> temp = stack.pop();
            int count = temp.getValue();
            while (count-- > 0) {
                res.append(temp.getKey());
            }
        }
        return res.reverse().toString();
    }
}

File: Algorithms/1209_Remove_All_Adjacent_Duplicates_in_String_II/Python/Remove_All_Adjacent_Duplicates_in_String_II_Solution_2.py

# Space: O(n)
# Time: O(n)

class Solution:
    def removeDuplicates(self, s, k):
        length = len(s)
        if length <= 1: return s
        if k > length: return s

        stack = [['#', 1]]
        for char in s:
            if char == stack[-1][0]:
                stack[-1][1] += 1
            else:
                stack.append([char, 1])

            while stack[-1][1] >= k:
                stack[-1][1] -= k
                if stack[-1][1] == 0: stack.pop()

        return ''.join([i[0] * i[1] for i in stack[1:]])













